src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,596 @@
+                        return true;
+                    } else {
+                        var full = List.copyOf(window);
+                        window.clear();
+                        return downstream.push(full);
+                    }
+                }),
+
+                // Finisher
+                (window, downstream) -> {
+                    if(!downstream.isRejecting() && !window.isEmpty())
+                        downstream.push(List.copyOf(window));
+                }
+        );
+    }
+
+    /**
+     * Gathers elements into sliding windows, sliding out the most previous
+     * element and sliding in the next element for each subsequent window.
+     * If the stream is empty then no window will be produced. If the size of
+     * the stream is smaller than the window size then only one window will
+     * be emitted, containing all elements.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: [[1, 2], [2, 3], [3, 4], [4, 5], [5, 6], [6, 7], [7, 8]]
+     * List<List<Integer>> windows =
+     *     Stream.of(1,2,3,4,5,6,7,8).gather(Gatherers.windowSliding(2)).toList();
+     * }
+     *
+     * @param windowSize the size of the windows
+     * @param <TR> the type of elements the returned gatherer consumes
+     *             and the contents of the windows it produces
+     * @return a new gatherer which groups elements into sliding windows
+     * @throws IllegalArgumentException when windowSize is less than 1
+     */
+    public static <TR> Gatherer<TR, ?, List<TR>> windowSliding(int windowSize) {
+        if (windowSize < 1)
+            throw new IllegalArgumentException("'groupSize' must be greater than zero");
+
+        @SuppressWarnings("serial")
+        class SlidingWindow extends ArrayDeque<TR> {
+            boolean firstWindow = true;
+        }
+        return Gatherer.ofSequential(
+                // Initializer
+                SlidingWindow::new,
+
+                // Integrator
+                Integrator.ofGreedy((window, e, downstream) -> {
+                    window.addLast(e);
+                    if (window.size() < windowSize) {
+                        return true;
+                    } else {
+                        var full = List.copyOf(window);
+                        window.removeFirst();
+                        window.firstWindow = false;
+                        return downstream.push(full);
+                    }
+                }),
+
+                // Finisher
+                (window, downstream) -> {
+                    if(window.firstWindow && !downstream.isRejecting() && !window.isEmpty())
+                        downstream.push(List.copyOf(window));
+                }
+        );
+    }
+
+    /**
+     * An operation which performs an ordered, <i>reduction-like</i>,
+     * transformation for scenarios where no combiner-function can be
+     * implemented, or for reductions which are intrinsically
+     * order-dependent.
```

Comment:
I think, we should highlight the most important difference between `fold` and `reduce`: `reduce` must provide an associative function, while `fold` does not require this restriction.

Reply:
Please take a look here, probably this comment was overlooked?

Reply:
Yeah, I was thinking about that. I tried to word it will "less jargon" so it is easier to casually read and understand.
I'm wondering if adding "associativity" helps in that regard? 🤔